feat(py,ts): add v0.6 displacement field transform support (#575)
thewtex/ngff-zarr: docs/python.md

scale according the OME-Zarr data model. Note that the correct `scale` and
`translation` for each scale are automatically computed.

## Displacement and coordinate fields (v0.6)

OME-Zarr v0.6 (RFC-5) stores displacement and coordinate fields as ordinary
multiscale images. The only difference from a regular image is that the
vector-component axis, in the channel position, carries `type="displacement"`

> (or `type="coordinate"`) instead of `type="channel"`. Set it with the

Reviewer: Also should mention that this coordinate or displacement dimension is `discrete` like a `c` / channel / component dimension.

`axes_types` argument of [`to_multiscales`], which maps a dimension name to its
axis type:

```python
import dask.array as da
import numpy as np
import ngff_zarr as nz

# A 2-component displacement field over a yx image: the "c" dimension holds the
# (y, x) displacement vector, one component per output spatial axis.
data = da.from_array(np.zeros((2, 256, 256), dtype=np.float32))
field = nz.NgffImage(
    data=data,
    dims=("c", "y", "x"),
    scale={"c": 1.0, "y": 1.0, "x": 1.0},
    translation={"c": 0.0, "y": 0.0, "x": 0.0},
)

multiscales = nz.to_multiscales(field, axes_types={"c": "displacement"})

> nz.to_ngff_zarr("displacement.ome.zarr", multiscales, version="0.6")

Reviewer: Instead, and with implementing RFC-3 in mind, we want this to instead change the v06 definition of AxesType:

https://github.com/fideus-labs/ngff-zarr/blob/e02679a347b51aba8fd44f401ea7ef13bb64684d/py/ngff_zarr/v04/zarr_metadata.py#L27

to extend the union to include Literal["displacement"] and Literal["coordinate"].

And add `axes_types` to NgffImage here:

https://github.com/fideus-labs/ngff-zarr/blob/main/py/ngff_zarr/ngff_image.py#L22

Then specify with the constructor to `NgffImage` with `axes_types={"c": "displacement"}`

For more detailed examples and advanced usage, see the
[HCS documentation](./hcs.md).

> ## v0.6 Field Transforms (PR 145)

Reviewer: Remove identification of "PR 145" throughout the docs, implementation, tests, filenames. This will be part of OME-Zarr v0.6 and PR 145 is not meaningful to the reader.

Author: Addressed in the amended commit. The user-facing docs no longer identify this as PR 145; the wording now describes the feature as OME-Zarr v0.6 field transforms.

OME-Zarr v0.6 adds `coordinates` and `displacements` transforms that point at
a separate displacement-field multiscale group. The field data lives in that
group; the vector-component axis is a discrete `displacement` (or `coordinate`)

> axis followed by the spatial axes. When applying registration, always use the

Reviewer: ```suggestion
axis followed by the spatial axes. When resampling for analysis, prefer the
```

> finest scale level (index 0); coarser levels are for quick visualization only.

Reviewer: ```suggestion
finest scale level (index 0); coarser levels are usually for quick visualization.
```

```python
from ngff_zarr.v06.zarr_metadata import (
    Axis,
    CoordinateSystem,
    CoordinateSystemIdentifier,
    Dataset,
    Displacements,
    Metadata,
    Scale,
    Translation,
    TransformSequence,
)

# Spatial axes shared by both coordinate systems
spatial_axes = [
    Axis(name="y", type="space", unit="micrometer"),
    Axis(name="x", type="space", unit="micrometer"),
]

# Image multiscale: a top-level displacements transform maps
# the intrinsic image space to a registered output space.
image_metadata = Metadata(
    name="registered-image",
    coordinateSystems=[
        CoordinateSystem(name="intrinsic", axes=spatial_axes),
        CoordinateSystem(name="registered", axes=spatial_axes),
    ],
    datasets=[
        Dataset(

>             path="0",

Reviewer: ```suggestion
            path="scale0",
```

            coordinateTransformations=[
                TransformSequence(
                    name="scale0_to_intrinsic",
                    input=CoordinateSystemIdentifier(path="0"),
                    output=CoordinateSystemIdentifier(name="intrinsic"),
                    transformations=[
                        Scale(scale=[1.0, 1.0]),

>                         Translation(translation=[0.0, 0.0]),

Reviewer: We are duplicating state with the `NgffImage` constructor below. The `NgffImage` constructor should come first. Then we update the coordinateTransformations with Displacements after.

Author: Addressed in the amended commit. The Python docs now construct the `NgffImage` first, build the multiscale metadata around it, and attach the top-level `Displacements` transform after the `NgffMultiscales` object exists.

    coordinateSystems=[CoordinateSystem(name="intrinsic", axes=field_axes)],
    datasets=[
        Dataset(
            path="0",
            coordinateTransformations=[
                TransformSequence(
                    name="scale0_to_intrinsic",
                    input=CoordinateSystemIdentifier(path="0"),
                    output=CoordinateSystemIdentifier(name="intrinsic"),
                    transformations=[
                        Scale(scale=[1.0, 1.0, 1.0]),
                        Translation(translation=[0.0, 0.0, 0.0]),
                    ],
                )
            ],
        )
    ],
)
```

To write these as an OME-Zarr v0.6 store, pass `version='0.6'` to
[`to_ngff_zarr`]:

```python
import dask.array as da
import numpy as np
import ngff_zarr as nz
from ngff_zarr import NgffImage, NgffMultiscales

shape = (4, 5)
image = NgffImage(
    data=da.from_array(np.zeros(shape, dtype=np.float32), chunks=shape),
    dims=("y", "x"),
    scale={"y": 1.0, "x": 1.0},
    translation={"y": 0.0, "x": 0.0},
)
multiscales = NgffMultiscales(images=[image], metadata=image_metadata)
nz.to_ngff_zarr("registered.ome.zarr", multiscales, version="0.6")
```

> 

Author: Addressed in the amended commit. The Python example now writes both the image multiscale and the referenced displacement-field multiscale at `registered.ome.zarr/coordinateTransformations/displacementField`, so the transform path exists in the output store.

axis followed by the spatial axes. When resampling for analysis, prefer the
finest scale level (index 0); coarser levels are usually for quick visualization.

```python
import dask.array as da
import numpy as np
import ngff_zarr as nz
from ngff_zarr import NgffImage, NgffMultiscales
from ngff_zarr.v06.zarr_metadata import (
    Axis,
    CoordinateSystem,
    CoordinateSystemIdentifier,
    Dataset,
    Displacements,
    Metadata,
    Scale,
    Translation,
    TransformSequence,
)

shape = (4, 5)
image = NgffImage(
    data=da.from_array(np.zeros(shape, dtype=np.float32), chunks=shape),
    dims=("y", "x"),
    scale={"y": 1.0, "x": 1.0},
    translation={"y": 0.0, "x": 0.0},
)

# Spatial axes shared by both coordinate systems
spatial_axes = [
    Axis(name="y", type="space", unit="micrometer"),
    Axis(name="x", type="space", unit="micrometer"),
]

# Image multiscale metadata starts with the coordinate systems and dataset
# transform derived from the image.
image_metadata = Metadata(
    name="registered-image",
    coordinateSystems=[
        CoordinateSystem(name="intrinsic", axes=spatial_axes),
        CoordinateSystem(name="registered", axes=spatial_axes),

>     ],

Reviewer: Before this, let's create the coordinate system instances and re-use them, i.e.

```
intrinsic_coordinate_system = CoordinateSystem(name="intrinsic", axes=spatial_axes)
[...]
```

    datasets=[
        Dataset(

>             path="scale0",

Reviewer: ```suggestion
            path=image.name,
```

            coordinateTransformations=[
                TransformSequence(
                    name="scale0_to_intrinsic",
                    input=CoordinateSystemIdentifier(path="scale0"),
                    output=CoordinateSystemIdentifier(name="intrinsic"),
                    transformations=[
                        Scale(scale=[1.0, 1.0]),
                        Translation(translation=[0.0, 0.0]),
                    ],

>                 )

Reviewer: I think this should be generated by a utility function, e.g.

```python
def intrinsic_transform(image: NgffImage) -> TransformSequence:
    # generate a TransformSequence from the image.name, image.dims, image.scale, image.translation
    # name="intrinsic"
    # input coordinate system identifier path is "image.name"
```

@jo-mueller thoughts?

Reviewer 2: Tbh, the example seems a bit contrived to me? In practice , I think you'd only have to use the functionality around `NgffMultiscales` and `NgffImage` to create the displacement field vector zarr group. the *only* thing you really need to adjust, is that the displacement's axes types should be 

```json
{
  "name": "dfield",
  "axes": [
    {"name": "c", "type": "displacement"},
    {"name": "y", "type": "space"},
    {"name": "x", "type": "space"}
  ]
}
```

for a displacement transform on a yx image.

This could be done either post-hoc (edit `image.metadata` after instantiation?) or by exposing the axis' type as a new kwarg in the API (`axis_type`?).

Reviewer: This is more about creating the metadata for OME-Zarr write in a compatible, ergonomic way that does not duplicate state.
